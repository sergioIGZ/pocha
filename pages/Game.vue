<template>
  <div>
    <span v-if="!isGameStarted">Nueva partida</span>
    <div v-else>
      <div class="flex justify-center">
        <div>Ronda:</div>
        <div class="font-bold">{{ $store.state.currentGame.round }}</div>
      </div>
      <div class="flex justify-center">
        Carta/s:
        <div class="font-bold">{{ cards }}</div>
      </div>
    </div>
    <button v-if="!isGameStarted" @click="startGame">Empezar</button>
    <template v-else>
      <div class="flex justify-center">
        <button v-show="round > 1" @click="previousRound" class="bg-yellow-500">
          <span class="icon-loop" />
        </button>
        <button @click="validateRound">
          Siguiente Ronda >
        </button>
        <button
          v-if="!isEditing"
          v-show="round > 1"
          @click="isEditing = true"
          class="bg-green-500"
        >
          <span class="icon-pencil" />
        </button>
      </div>
    </template>
    <div class="middle">
      <players-list
        v-if="!isGameStarted"
        :players="gamePlayerNames"
        :show-delete="!isGameStarted"
        @on-delete-player="deleteGamePlayer"
        class="marked"
      />
      <points-player-list
        v-else
        :is-editing="isEditing"
        :players="gamePlayers"
        :max-points="cards"
        @set-player-bet="setPlayerBet"
        @set-player-points="setPlayerPoints"
        @cancel-edit="isEditing = false"
        @save-edit="saveEditPoints"
        class="marked"
      />
      <div v-if="showNotes" class="flex flex-col p-2">
        <label for="notes">Notas</label>
        <textarea id="notes" v-model="notes" rows="3" />
      </div>
    </div>
    <players-list
      v-if="showSelectPlayers"
      :players="$store.getters.playersNotInGame"
      :show-delete="false"
      @on-player-click="choosePlayer"
    />
    <div class="bottom">
      <button
        v-if="!isGameStarted"
        @click="showSelectPlayers = !showSelectPlayers"
      >
        + Jugador
      </button>
      <button @click="$router.push('/')" class="bg-red-500">Salir</button>
      <button @click="saveGame" class="bg-green-500">Guardar</button>
      <button @click="toggleNotes" class="bg-green-500">Notas</button>
    </div>
  </div>
</template>

<script>
import PlayersList from '../components/playersList/playersList'
import PointsPlayerList from '../components/pointsPlayerList/pointsPlayerList'
export default {
  name: 'Game',
  components: { PlayersList, PointsPlayerList },
  data() {
    return {
      showSelectPlayers: false,
      isEditing: false,
      showNotes: false,
      notes: ''
    }
  },
  computed: {
    gamePlayerNames() {
      return this.$store.state.currentGame.players.map(player => player.name)
    },
    gamePlayers() {
      return this.$store.state.currentGame.players
    },
    round() {
      return this.$store.state.currentGame.round
    },
    isGameStarted() {
      return this.$store.state.currentGame.round !== 0
    },
    cards() {
      const numberOfPlayers = this.gamePlayers.length
      const coefficient = Math.trunc(40 / numberOfPlayers)
      const round = this.round
      let cards = 0

      if (
        round <= numberOfPlayers ||
        round > 2 * numberOfPlayers + 2 * (coefficient - 2)
      ) {
        cards = 1
      } else if (
        round > numberOfPlayers &&
        round < numberOfPlayers + coefficient - 1
      ) {
        cards = round - numberOfPlayers + 1
      } else if (
        round >= numberOfPlayers + coefficient - 1 &&
        round <= 2 * numberOfPlayers + coefficient - 2
      ) {
        cards = coefficient
      } else if (
        round >= numberOfPlayers + coefficient - 1 &&
        round < 2 * numberOfPlayers + 2 * (coefficient - 2) + 1
      ) {
        cards = 2 * numberOfPlayers + coefficient - 2 + coefficient - round
      } else {
        cards = 9999999
      }
      return cards
    }
  },
  methods: {
    startGame() {
      if (this.gamePlayers.length < 3) {
        return this.$store.commit('showAlert', {
          isVisible: true,
          type: 'info',
          title: 'A veeer...',
          message: 'No hay jugadores suficientes'
        })
      } else {
        this.$store.commit('startGame')
      }
    },
    deleteGamePlayer(playerName) {
      this.$store.commit('deleteGamePlayer', playerName)
    },
    choosePlayer(playerName) {
      this.$store.commit('choosePlayer', playerName)
      this.showSelectPlayers = false
    },
    setPlayerBet({ name, bet }) {
      this.$store.commit('setPlayerBet', { name, bet })
    },
    setPlayerPoints({ name, points }) {
      this.$store.commit('setPlayerPoints', { name, points })
    },
    doNextRound() {
      this.$store.commit('nextRound')
    },
    previousRound() {
      this.$store.commit('showAlert', {
        isVisible: true,
        type: 'info',
        title: 'Ojo!',
        message:
          'Estás seguro ke kieres viajar en el tiempo a la ronda anterior? acaso alguien la ha cagado?',
        yesNoButtons: true,
        yesAction: () => this.$store.commit('resetRound')
      })
    },
    validateRound() {
      const cards = this.cards
      const playerBets = this.gamePlayers.reduce(
        (acc, next) => acc + next.bet,
        0
      )
      const playerPoints = this.gamePlayers.reduce(
        (acc, next) => acc + next.points,
        0
      )
      if (playerBets === cards) {
        return this.$store.commit('showAlert', {
          isVisible: true,
          type: 'error',
          title: 'Ojo!',
          message:
            'Las apuestas son igual a las cartas. Alguien tiene ke pringar'
        })
      }
      if (playerPoints !== cards) {
        return this.$store.commit('showAlert', {
          isVisible: true,
          type: 'error',
          title: 'Ojo!',
          message:
            'Las jugadas ganadas tienen que ser igual al numero de cartas repartidas. Te has colao bacalao'
        })
      }

      this.doNextRound()
    },
    toggleNotes() {
      this.showNotes = !this.showNotes
    },
    saveEditPoints(playerPoints) {
      this.$store.commit('setPlayerTotalPoints', playerPoints)
      this.isEditing = false
    },
    saveGame() {
      this.$store.commit('showAlert', {
        isVisible: true,
        type: 'info',
        title: 'Seguro??',
        message:
          'Estás seguro de ke kieres guardar la puntuación? asegurate de ke no la hayas guardado antes pa ke no se repita',
        yesNoButtons: true,
        yesAction: () => this.$store.dispatch('saveGame')
      })
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.marked {
  color: #42b983;
}
.bottom {
  display: flex;
  justify-content: space-around;
  align-items: center;
  position: fixed;
  bottom: 36px;
}
</style>
