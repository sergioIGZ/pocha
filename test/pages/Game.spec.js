import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Game from '@/pages/Game'

const localVue = createLocalVue()
localVue.use(Vuex)

const mockStore = ({ round = 1, players = [] } = {}) => {
  return new Vuex.Store({
    state: () => ({
      players: [],
      currentGame: {
        players,
        round
      },
      alert: {
        isVisible: false,
        type: null,
        message: '',
        title: ''
      }
    }),
    mutations: {
      showAlert: jest.fn(),
      nextRound: jest.fn()
    },
    getters: {
      playersNotInGame: jest.fn(() => [])
    }
  })
}

const fourPlayers = [
  { name: 'foo1' },
  { name: 'foo2' },
  { name: 'foo3' },
  { name: 'foo4' }
]
const tenPlayers = [
  { name: 'foo1' },
  { name: 'foo2' },
  { name: 'foo3' },
  { name: 'foo4' },
  { name: 'foo5' },
  { name: 'foo6' },
  { name: 'foo7' },
  { name: 'foo8' },
  { name: 'foo9' },
  { name: 'foo10' }
]

describe('Game Page', () => {
  describe('Behaviour', () => {
    it('should be mounted with no problems', () => {
      // Arrange.
      // Act.
      const wrapper = shallowMount(Game, {
        localVue,
        store: mockStore
      })
      // Assert.
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Methods', () => {
    describe('doNextRound', () => {
      describe('cards in round', () => {
        it.each`
          round | nextRound | players        | numberOfPlayers       | expectedCards
          ${1}  | ${2}      | ${fourPlayers} | ${fourPlayers.length} | ${1}
          ${3}  | ${4}      | ${fourPlayers} | ${fourPlayers.length} | ${1}
          ${5}  | ${6}      | ${fourPlayers} | ${fourPlayers.length} | ${2}
          ${12} | ${13}     | ${fourPlayers} | ${fourPlayers.length} | ${9}
          ${13} | ${14}     | ${fourPlayers} | ${fourPlayers.length} | ${10}
          ${16} | ${17}     | ${fourPlayers} | ${fourPlayers.length} | ${10}
          ${17} | ${18}     | ${fourPlayers} | ${fourPlayers.length} | ${9}
          ${24} | ${26}     | ${fourPlayers} | ${fourPlayers.length} | ${2}
          ${25} | ${26}     | ${fourPlayers} | ${fourPlayers.length} | ${1}
          ${1}  | ${2}      | ${tenPlayers}  | ${tenPlayers.length}  | ${1}
          ${9}  | ${10}     | ${tenPlayers}  | ${tenPlayers.length}  | ${1}
          ${11} | ${12}     | ${tenPlayers}  | ${tenPlayers.length}  | ${2}
          ${12} | ${13}     | ${tenPlayers}  | ${tenPlayers.length}  | ${3}
          ${13} | ${14}     | ${tenPlayers}  | ${tenPlayers.length}  | ${4}
          ${22} | ${23}     | ${tenPlayers}  | ${tenPlayers.length}  | ${4}
          ${23} | ${24}     | ${tenPlayers}  | ${tenPlayers.length}  | ${3}
          ${24} | ${25}     | ${tenPlayers}  | ${tenPlayers.length}  | ${2}
          ${25} | ${26}     | ${tenPlayers}  | ${tenPlayers.length}  | ${1}
        `(
          'should set cards = $expectedCards when is the $round (passing to the $nextRound) round and we have $numberOfPlayers',
          ({ round, players, expectedCards }) => {
            // Arrange.
            const wrapper = shallowMount(Game, {
              localVue,
              store: mockStore({ round, players })
            })

            // Act.
            wrapper.vm.doNextRound()

            // Assert.
            expect(wrapper.vm.cards).toBe(expectedCards)
          }
        )
      })
    })
  })
})
