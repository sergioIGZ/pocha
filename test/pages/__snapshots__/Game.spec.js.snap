// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Page Behaviour should be mounted with no problems 1`] = `
<div>
  <div><span>Ronda: 1</span> <span>Carta/s: 2</span></div>
  <div class="flex justify-center"><button class="bg-yellow-500" style="display: none;"><span class="icon-loop"></span></button> <button>
      Siguiente Ronda &gt;
    </button> <button class="bg-green-500" style="display: none;"><span class="icon-pencil"></span></button></div>
  <div class="middle">
    <points-player-list-stub players="" maxpoints="2" showdelete="true" class="marked"></points-player-list-stub>
    <!---->
  </div>
  <!---->
  <div class="bottom">
    <!----> <button class="bg-red-500">Salir</button> <button class="bg-green-500">Notas</button></div>
</div>
`;
